File: android-sdk/src/main/java/io/tezos/mobile/rpc/TezosRpcClient.kt
package io.tezos.mobile.rpc

import okhttp3.Call
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody

class TezosRpcClient(
    private val baseUrl: String,
    private val httpClient: Call.Factory = OkHttpClient()
) {
    private val jsonMedia = "application/json".toMediaType()
    init {
        require(baseUrl.lowercase().startsWith("https://")) { "TezosRpcClient requires an HTTPS baseUrl" }
    }
    @Throws(Exception::class)
    fun getHeadHash(): String {
        val request = Request.Builder()
            .url("$baseUrl/chains/main/blocks/head/hash")
            .get()
            .build()
        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw Exception("HTTP ${response.code}")
            return response.body?.string()?.trim('"', '\n', ' ', '\r') ?: throw Exception("Empty body")
        }
    }

    @Throws(Exception::class)
    fun getBalance(address: String): String {
        val request = Request.Builder()
            .url("$baseUrl/chains/main/blocks/head/context/contracts/$address/balance")
            .get()
            .build()
        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw Exception("HTTP ${response.code}")
            return response.body?.string()?.trim('"', '\n', ' ', '\r') ?: throw Exception("Empty body")
        }
    }

    @Throws(Exception::class)
    fun getCounter(address: String): String {
        val request = Request.Builder()
            .url("$baseUrl/chains/main/blocks/head/context/contracts/$address/counter")
            .get()
            .build()
        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw Exception("HTTP ${response.code}")
            return response.body?.string()?.trim('"', '\n', ' ', '\r') ?: throw Exception("Empty body")
        }
    }

    @Throws(Exception::class)
    fun getChainId(): String {
        val request = Request.Builder()
            .url("$baseUrl/chains/main/chain_id")
            .get()
            .build()
        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw Exception("HTTP ${response.code}")
            return response.body?.string()?.trim('"', '\n', ' ', '\r') ?: throw Exception("Empty body")
        }
    }

    @Throws(Exception::class)
    fun postRawString(path: String, jsonBody: String): String {
        val body = jsonBody.toRequestBody(jsonMedia)
        val request = Request.Builder()
            .url("$baseUrl$path")
            .post(body)
            .build()
        httpClient.newCall(request).execute().use { response ->
            if (!response.isSuccessful) throw Exception("HTTP ${response.code}")
            return response.body?.string()?.trim('"', '\n', ' ', '\r') ?: throw Exception("Empty body")
        }
    }

    @Throws(Exception::class)
    fun injectOperation(signedOperationHex: String): String {
        val json = "\"$signedOperationHex\""
        return postRawString("/injection/operation?chain=main", json)
    }
}

File: ios-sdk/Sources/TezosMobileSDK/RPC/TezosRPCClient.swift
import Foundation

public struct TezosRPCConfig {
    public let baseURL: URL
    public let urlSession: URLSession

    public init(baseURL: URL, urlSession: URLSession = .shared) {
        self.baseURL = baseURL
        self.urlSession = urlSession
    }
}

public enum TezosRPCError: Error {
    case invalidResponse
    case httpError(statusCode: Int, body: String?)
    case decodingError(underlying: Error)
}

public final class TezosRPCClient {
    private let config: TezosRPCConfig

    public init(config: TezosRPCConfig) {
        self.config = config
    }

    public func getHeadHash() async throws -> String {
        let path = "/chains/main/blocks/head/hash"
        let request = try makeRequest(path: path)
        let (data, response) = try await config.urlSession.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw TezosRPCError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8)
            throw TezosRPCError.httpError(statusCode: http.statusCode, body: body)
        }
        guard let hash = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) else {
            throw TezosRPCError.invalidResponse
        }
        return hash.replacingOccurrences(of: "\"", with: "")
    }

    public func getBalance(address: String) async throws -> String {
        let path = "/chains/main/blocks/head/context/contracts/\(address)/balance"
        let request = try makeRequest(path: path)
        let (data, response) = try await config.urlSession.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw TezosRPCError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8)
            throw TezosRPCError.httpError(statusCode: http.statusCode, body: body)
        }
        guard let balance = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines) else {
            throw TezosRPCError.invalidResponse
        }
        return balance.replacingOccurrences(of: "\"", with: "")
    }

    private func makeRequest(path: String) throws -> URLRequest {
        let url = config.baseURL.appendingPathComponent(path)
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        return request
    }
}


